internal/controller/apiextensions/composite/composition_functions.go

```diff
@@ -418,6 +425,29 @@ func (c *FunctionComposer) Compose(ctx context.Context, xr *composite.Unstructur
 	return CompositionResult{ConnectionDetails: d.GetComposite().GetConnectionDetails(), Composed: resources, Events: events}, nil
 }
 
+// The function `getComposedFieldOwnerName` generates a unique field owner name
+// for a given Crossplane composite resource (XR). This uniqueness is crucial to
+// prevent multiple XRs, which compose the same resource, from continuously
+// alternating as controllers.
+//
+// The function generates a deterministic hash based on the XR's name and
+// GroupKind (GK), ensuring consistency even during system restores. The hash
+// does not include the XR's UID (as it's not deterministic), namespace (XRs
+// don't have one), or version (to allow version changes without needing to
+// update the field owner name).
+//
+// We decided to include the GK in the hash to prevent transferring ownership of
+// composed resources across XRs with whole new GK, as that should not be
+// supported without manual intervention.
+//
+// Given that field owner names are limited to 128 characters, the function
+// truncates the hash to 32 characters. A longer hash was deemed unnecessary.
+func getComposedFieldOwnerName(xr *composite.Unstructured) string {
+	h := sha256.New()
+	_, _ = fmt.Fprintf(h, "%s-%s", xr.GetName(), xr.GroupVersionKind().GroupKind())
```

Comment:
```suggestion
	_, _ = h.Write([]byte(xr.GetName() + xr.GroupVersionKind().GroupKind()))
```

Nit: I don't think the `-` separator is needed here given we'll one-way hash it.